File: frameworks/dotnet/fastendpoints/UnitTests/ItemsTests.cs
namespace UnitTests;

public sealed class ItemsTests(App app) : TestBase<App>
{
    // rb:test items.read
    [Fact]
    [Trait("corpus", "items.read")]
    public async Task A_row_is_read_by_the_id_in_the_path()
    {
        using HttpResponseMessage response = await app.Client.GetAsync("/items/17", Cancellation);

        await Answer.IsAsync(Row(17), response);
    }

    // rb:test items.head
    [Fact]
    [Trait("corpus", "items.head")]
    public async Task Head_is_answered_by_the_read_endpoint()
    {
        // Kestrel leaves a HEAD answer's body unwritten. The test host does not, so the body is not checked.
        using HttpResponseMessage response = await app.Client.SendAsync(new(HttpMethod.Head, "/items/17"), Cancellation);

        Assert.Equal(HttpStatusCode.OK, response.StatusCode);
        Assert.Equal("application/json", response.Content.Headers.ContentType?.MediaType);
    }

    // rb:test items.create
    [Fact]
    [Trait("corpus", "items.create")]
    public async Task A_created_item_is_the_row_after_the_last_where_the_read_endpoint_finds_it()
    {
        using HttpResponseMessage response = await app.Client.PostAsync("/items", New(), Cancellation);

        Assert.Equal(HttpStatusCode.Created, response.StatusCode);
        Assert.Equal("/items/1426", response.Headers.Location?.OriginalString);
        await Answer.IsAsync(Numbered(1426), response);
    }

    // rb:test items.replace
    [Fact]
    [Trait("corpus", "items.replace")]
    public async Task A_replaced_item_takes_the_id_in_the_path()
    {
        using HttpResponseMessage response = await app.Client.PutAsync("/items/17", New(), Cancellation);

        await Answer.IsAsync(Numbered(17), response);
    }

    // rb:test items.update
    [Fact]
    [Trait("corpus", "items.update")]
    public async Task A_patch_is_merged_onto_the_row()
    {
        using HttpResponseMessage response = await app.Client.PatchAsync("/items/17", new ByteArrayContent(Expected.Bytes("items.patch.json"))
        {
            Headers = { ContentType = new("application/json") },
        }, Cancellation);

        JsonObject merged = Row(17).AsObject();
        foreach ((string name, JsonNode? value) in Expected.Json("items.patch.json").AsObject())
        {
            merged[name] = value?.DeepClone();
        }
        await Answer.IsAsync(merged, response);
    }

    // rb:test items.delete
    [Fact]
    [Trait("corpus", "items.delete")]
    public async Task A_delete_is_answered_204_with_no_body()
    {
        using HttpResponseMessage response = await app.Client.DeleteAsync("/items/17", Cancellation);

        Assert.Equal(HttpStatusCode.NoContent, response.StatusCode);
        Assert.Empty(await Answer.BytesAsync(response));
    }

    [Fact]
    public async Task An_id_that_is_not_an_integer_matches_no_route()
    {
        using HttpResponseMessage response = await app.Client.GetAsync("/items/seventeen", Cancellation);

        Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
    }

    private static JsonNode Row(int id) => Expected.Json("items.large.json")["items"]![id - 1]!.DeepClone();

    private static JsonObject Numbered(int id)
    {
        JsonObject item = new() { ["id"] = id };
        foreach ((string name, JsonNode? value) in Expected.Json("items.new.json").AsObject())
        {
            item[name] = value?.DeepClone();
        }
        return item;
    }

    private static ByteArrayContent New() => new(Expected.Bytes("items.new.json")) { Headers = { ContentType = new("application/json") } };
}

File: frameworks/java/spring-boot/UnitTests/src/test/java/unittests/ItemsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ObjectNode;

class ItemsTests extends SpringApp {

    // rb:test items.read
    @Test
    @Tag("items.read")
    void aRowIsReadByTheIdInThePath() throws Exception {
        Answer.is(Expected.row(17), get("/items/17"));
    }

    // rb:test items.head
    @Test
    @Tag("items.head")
    void headIsAnsweredByTheGetHandlerWithNoBody() throws Exception {
        HttpResponse<byte[]> response = send(request("/items/17").method("HEAD", HttpRequest.BodyPublishers.noBody()));

        assertEquals(200, response.statusCode());
        assertTrue(Answer.header(response, "content-type").startsWith("application/json"));
        assertEquals(0, response.body().length);
    }

    // rb:test items.create
    @Test
    @Tag("items.create")
    void aCreatedItemIsTheRowAfterTheLast() throws Exception {
        HttpResponse<byte[]> response = send("POST", "/items", "application/json", Expected.bytes("items.new.json"));

        assertEquals(201, response.statusCode());
        assertEquals("/items/1426", Answer.header(response, "location"));
        Answer.is(numbered(1426), response);
    }

    // rb:test items.replace
    @Test
    @Tag("items.replace")
    void aReplacedItemTakesTheIdInThePath() throws Exception {
        Answer.is(numbered(17), send("PUT", "/items/17", "application/json", Expected.bytes("items.new.json")));
    }

    // rb:test items.update
    @Test
    @Tag("items.update")
    void aPatchIsMergedOntoTheRow() throws Exception {
        HttpResponse<byte[]> response = send("PATCH", "/items/17", "application/json", Expected.bytes("items.patch.json"));

        ObjectNode merged = Expected.row(17);
        for (Map.Entry<String, JsonNode> field : Expected.json("items.patch.json").properties()) {
            merged.set(field.getKey(), field.getValue());
        }
        Answer.is(merged, response);
    }

    // rb:test items.delete
    @Test
    @Tag("items.delete")
    void aDeleteIsAnswered204WithNoBody() throws Exception {
        HttpResponse<byte[]> response = send(request("/items/17").DELETE());

        assertEquals(204, response.statusCode());
        assertEquals(0, response.body().length);
    }

    private static ObjectNode numbered(int id) {
        ObjectNode item = Expected.JSON.createObjectNode().put("id", id);
        item.setAll((ObjectNode) Expected.json("items.new.json"));
        return item;
    }
}
